### DISCUSSION TITLE
Duplicate Option Hints

### DISCUSSION THREAD

**Discussion by JasonLG1979** [librespot-org/librespot#782] [General]
While updating the Options section of the wiki I noticed several instances of duplicate option hints.

`PATH` is used twice, for both `cache` and `system-cache`.
`NAME` is used 3 times, for `name` (device name), `backend` and `mixer-name`
`TIME` is used twice, for both `normalisation-attack` and `normalisation-release`

I'd also like to discuss changing some of the arg and hint names so they make more sense. For example the hint for `volume-range` is just `RANGE` and for `normalisation-gain-type` it's just `TYPE` and so on. There are several that could be better.

**Comment by roderickvd** [comment]
Feel free to submit a PR and see where you can improve. I know you’re not a Rust programmer yet but this would be a great place to start. 

As for the hints, I made it that way. They are not unique variables, they are hints on what the values should look like. So it’s just as much a path to the cache as it is a path to the system cache. And a name for the device, just as much as a name for the mixer. And so on. 

Along these same lines, I _do_ agree with the range and type. I thought I had already done this but I maybe I made a mistake resolving conflicts.

**Reply by JasonLG1979** [reply]
> Feel free to submit a PR and see where you can improve. I know you’re not a Rust programmer but this would be a great place to start.

~~I'll give it a shot.~~ No point.

> As for the hints, I made it that way. They are not unique variables, they are hints on what the values should look like. So it’s just as much a path to the cache as it is a path to the system cache. And a name for the device, just as much as a name for the mixer. And so on.

~~The hints can be used as environment variables. They have to be unique or they are useless as environment variables.~~ Nope. My whole premise was incorrect. 

> Along these same lines, I do agree with the range and type. I thought I had already done this but I maybe I made a mistake resolving conflicts.

~~Before I start I'll post a proposed list here. The options should not be to hard the parsing of bools might take a bit. I'll likely do it in 2 PR's.~~

**Reply by roderickvd** [reply]
I considered it, but I don't think it does. Not on my Mac or RPi as environment variables anyway.

**Reply by JasonLG1979** [reply]
> I considered it, but I don't think it does. Not on my Mac or RPi as environment variables anyway.

~~It does. Look how [raspotify](https://github.com/dtcooper/raspotify) uses environment vars in it's systemd unit file and an environment file as a config.~~ Nope sure doesn't...

**Reply by JasonLG1979** [reply]
Here are my proposed Option Hints. I would also like to change a few option names to make them more clear. Basically make them align with my proposed hints. some are not as clear as they could be IMHO. For example `mixer` is a horrible name. It should really be `volume-type`.  
| Option                  | Hint                         |
|-------------------------|------------------------------|
| cache                   | CACHE_PATH                   |
| system-cache            | SYSTEM_CACHE_PATH            |
| cache-size-limit        | CACHE_SIZE_LIMIT             |
| name                    | DEVICE_NAME                  |
| device-type             | DEVICE_TYPE                  |
| bitrate                 | BITRATE                      |
| onevent                 | ONEVENT_PROGRAM              |
| username                | USERNAME                     |
| password                | PASSWORD                     |
| proxy                   | PROXY_URL                    |
| ap-port                 | AP_PORT                      |
| backend                 | BACKEND_NAME                 |
| device                  | AUDIO_DEVICE_NAME            |
| format                  | PCM_FORMAT                   |
| dither                  | DITHER_TYPE                  |
| mixer                   | VOLUME_TYPE                  |
| mixer-name              | VOLUME_MIXER_NAME            |
| mixer-card              | VOLUME_MIXER_CARD            |
| mixer-index             | VOLUME_MIXER_INDEX           |
| initial-volume          | INITIAL_VOLUME               |
| zeroconf-port           | ZEROCONF_PORT                |
| normalisation-method    | NORMALISATION_METHOD         |
| normalisation-gain-type | NORMALISATION_GAIN_TYPE      |
| normalisation-pregain   | NORMALISATION_PREGAIN_DB     |
| normalisation-threshold | NORMALISATION_THRESHOLD_DBFS |
| normalisation-attack    | NORMALISATION_ATTACK_TIME    |
| normalisation-release   | NORMALISATION_RELEASE_TIME   |
| normalisation-knee      | NORMALISATION_KNEE_STEEPNESS |
| volume-ctrl             | VOLUME_CTRL_SCALE            |
| volume-range            | VOLUME_RANGE                 |

**Reply by roderickvd** [reply]
Not by that virtue: https://github.com/dtcooper/raspotify/blob/629bcd1b39b00bf3cc27463c83e32b43c29b635b/raspotify/lib/systemd/system/raspotify.service#L19

`getopts` doesn't check if hints are set as environment variables. `raspotify` passes self-named variables on the command line that happen to align with `librespot` argument names. They might as well have been called FOO and BAR. I also like WALDO.

**Reply by JasonLG1979** [reply]
> `getopts` doesn't check if hints are set as environment variables. `raspotify` passes self-named variables on the command line that happen to align with `librespot` argument names. They might as well have been called FOO and BAR. I also like WALDO.

Well then I'm an idiot... Checking environment variables would be nice though.

**Reply by JasonLG1979** [reply]
Never mind I guess.

**Reply by roderickvd** [reply]
Don't fret, might as well have been me who missed it. There does not seem to be a definite standard on how to format `getopts`. Best references I could find are in #757.

I hesitated renaming option names, to not stir up too much trouble with downstream distributions like moOde. But since we might as well be calling `0.3.0` "Breaking Bad" :laughing: we can think about doing it.

**Reply by JasonLG1979** [reply]
@roderickvd You can close this or whatever you do to discussions as my whole premise is flawed.

**Reply by JasonLG1979** [reply]
> I hesitated renaming option names, to not stir up too much trouble with downstream distributions like moOde. But since we might as well be calling `0.3.0` "Breaking Bad"  we can think about doing it.

If the hints truly mean nothing then what the hell are they even good for? I mean you could use them to reference the argument's expected type. String, Bool and so on but there's already descriptions so they seem rather redundant.

**Reply by JasonLG1979** [reply]
As far as arg changes I propose these to make the args more clear and self-explanatory:

| Current                 | Proposed                     |
|-------------------------|------------------------------|
| name                    | device-name                  |
| device                  | audio-device-name            |
| format                  | pcm-format                   |
| dither                  | dither-type                  |
| mixer                   | volume-type                  |
| mixer-name              | volume-mixer-name            |
| mixer-card              | volume-mixer-card            |
| mixer-index             | volume-mixer-index           |
| volume-ctrl             | volume-ctrl-scale            | 

The mixer card and index args could really be combined I'm pretty sure. Much like device is can be used to specify the output device, `hw:0,0` for example.

**Reply by JasonLG1979** [reply]
As far as the hints I propose a noun type pair when possible. So `device-name` would be `NAME_STRING`, `ap-port` would be `PORT_INT` and so on.

**Reply by JasonLG1979** [reply]
Something like this as far as giving the hints some meaning and use:

| Option                  | Hint                         |
|-------------------------|------------------------------|
| cache                   | PATH_STRING                  |
| system-cache            | PATH_STRING                  |
| cache-size-limit        | SIZE_STRING                  |
| name                    | NAME_STRING                  |
| device-type             | TYPE_STRING                  |
| bitrate                 | BITRATE_INTEGER              |
| onevent                 | PROGRAM_STRING               |
| username                | USERNAME_STRING              |
| password                | PASSWORD_STRING              |
| proxy                   | URL_STRING                   |
| ap-port                 | PORT_INTEGER                 |
| backend                 | NAME_STRING                  |
| device                  | NAME_STRING                  |
| format                  | FORMAT_STRING                |
| dither                  | TYPE_STRING                  |
| mixer                   | TYPE_STRING                  |
| mixer-name              | NAME_STRING                  |
| mixer-card              | CARD_STRING                  |
| mixer-index             | INDEX_INTEGER                |
| initial-volume          | PERCENT_INTEGER              |
| zeroconf-port           | PORT_INTEGER                 |
| normalisation-method    | METHOD_STRING                |
| normalisation-gain-type | TYPE_STRING                  |
| normalisation-pregain   | DB_FLOAT                     |
| normalisation-threshold | DBFS_FLOAT                   |
| normalisation-attack    | TIME_INTEGER                 |
| normalisation-release   | TIME_INTEGER                 |
| normalisation-knee      | STEEPNESS_FLOAT              |
| volume-ctrl             | TYPE_STRING                  |
| volume-range            | DB_FLOAT                     |

**Comment by roderickvd** [comment]
This was continued in #784.
